File: 2GII/PDOO/QytetetRuby/lib/jugador.rb
#encoding: utf-8

require_relative "casilla"
require_relative "titulo_propiedad"
require_relative "sorpresa"
require_relative "qytetet"
require_relative "especulador"

module ModeloQytetet
  
  class Jugador
    attr_reader :nombre, :saldo, :propiedades
    attr_writer :carta_libertad
    attr_accessor :encarcelado, :casilla_actual
    

    # -------------------------- Constructor -----------------------------------    
    def initialize(nombre)
      @nombre = nombre
      @encarcelado = false
      @saldo = 7_500
      @propiedades = nil
      @casilla_actual = Casilla.new(0, TipoCasilla::SALIDA)    
      
      @@factor_especulador = 1
    end

	def nuevoJugador(jugador)
    @nombre = jugador.nombre
    @encarcelado = jugador.encarcelado
    @saldo = jugador.saldo
    @propiedades = jugador.propiedades
    @carta_libertad = jugador.get_carta_libertad
    @casilla_actual = jugador.casilla_actual	  
	end
    
    # ----------------------- Consultores -------------------------------------   
    def get_carta_libertad()
      @carta_libertad
    end
    
    def obtener_capital
      capital = @saldo
      n_casas = n_hoteles = 0
      
      if @propiedades != nil
        @propiedades.each do |propiedad|  
          n_casas = propiedad.casilla.num_casas
          n_hoteles = propiedad.casilla.num_hoteles
          capital += propiedad.factor_revalorizacion * (n_casas + n_hoteles) + \
                     propiedad.casilla.coste
        end        
      end      
      
      capital
    end

    def tengo_propiedades
      (@propiedades == nil) ? false : true 
    end
    
    def tengo_carta_libertad
      (@carta_libertad == nil) ? false : true
    end   
        
    def tengo_saldo(cantidad)
      @saldo >= cantidad
    end    
    
    def puedo_edificar_casa(casilla)
      es_de_mi_propiedad(casilla) && tengo_saldo(casilla.titulo.precio_edificar)
    end
    
    def puedo_edificar_hotel(casilla)
      es_de_mi_propiedad(casilla) && tengo_saldo(casilla.titulo.precio_edificar)
    end
    
    def puedo_hipotecar(casilla)
      !casilla.esta_hipotecada
    end
    
    def puedo_pagar_hipoteca(casilla)
      precio_deshipotecar = casilla.get_coste_cancelar_hipoteca
      
      tengo_saldo(precio_deshipotecar) ? true : false
    end
    
    def puedo_vender_propiedad(casilla)
      es_de_mi_propiedad(casilla) && !casilla.titulo.hipotecada
    end      
    
    def cuantas_casas_hoteles_tengo
      total = 0
      
      if @propiedades != nil 
        @propiedades.each do |propiedad|
          cas = propiedad.casilla
          total += cas.num_casas + cas.num_hoteles
        end
      end 
      
      total
    end

    # Comprueba si está el objeto
    def es_de_mi_propiedad(casilla)
      @propiedades.include?(casilla.titulo)
    end
    
    # ----------------------- Modificadores -----------------------------------   
    def modificar_saldo(cantidad)
      @saldo += cantidad
    end    
    
    # Borra todas las ocurrencias, en este caso como máximo solo una
    def eliminar_de_mis_propiedades(casilla)
      @propiedades.delete(casilla.titulo)
    end
    
    # ---------------------------- Métodos -------------------------------------    
    def actualizar_posicion(casilla)
      tiene_propietario = false
      
      if(casilla.numero_casilla < @casilla_actual.numero_casilla)
        modificar_saldo(Qytetet.class_variable_get(:@@SALDO_SALIDA))
      end
      
      @casilla_actual = casilla
      
      if (casilla.soy_edificable)
        tiene_propietario = casilla.tengo_propietario
        if (tiene_propietario && !casilla.propietario_encarcelado)
          coste_alquiler = casilla.cobrar_alquiler()
          modificar_saldo(-coste_alquiler)
        end      
      elsif (casilla.tipo == TipoCasilla::IMPUESTO)
        pagar_impuestos(-casilla.valor)
      end
      
      tiene_propietario
    end
    
    def comprar_titulo
      puedo_comprar = false
      
      if (@casilla_actual.soy_edificable && !@casilla_actual.tengo_propietario\
                                         && @casilla_actual.coste <= @saldo)
        if @propiedades == nil
          @propiedades = Array.new
        end
        
        @propiedades << @casilla_actual.asignar_propietario(self)                             
        modificar_saldo(-@casilla_actual.coste)
        puedo_comprar = true
      end
      
      puedo_comprar
    end
    
    def devolver_carta_libertad
      a_devolver = @carta_libertad
      @carta_libertad = nil
      
      a_devolver
    end
    
    # Crash or not to crash ? That is the question
    def ir_a_carcel(casilla = Casilla.new(15, TipoCasilla::CARCEL))
      @casilla_actual = casilla
      @encarcelado = true
    end        

    def obtener_propiedades_hipotecadas(hipotecada)      
      prop_hipotecadas = Array.new
      
      @propiedades.each do |propiedad|  
        if (propiedad.hipotecada == hipotecada)
          prop_hipotecadas << propiedad
        end
      end
    end
    
    def pagar_cobrar_por_casa_y_hotel(cantidad)
      modificar_saldo(cantidad * cuantas_casas_hoteles_tengo)
    end
    
    def pagar_libertad(cantidad = Qytetet.class_variable_get(:@@PRECIO_LIBERTAD))
      pagado = false
      
      if (tengo_saldo(cantidad))
        modificar_saldo(-cantidad)
        pagado = true
      end
      
      pagado
    end
         
    def vender_propiedad(casilla)
      modificar_saldo(casilla.vender_titulo)
      
      eliminar_de_mis_propiedades(casilla)
      if @propiedades.empty?
        @propiedades = nil
      end
    end        
    
    def pagar_impuestos(cantidad)
      modificar_saldo(-cantidad)
    end
    
    def convertirme(fianza)
      Especulador.new(self, fianza)
    end
    
    def to_s()
      (@encarcelado == true)? encarcelado = "Si" : encarcelado = "No"
      
      props = String.new    
      if @propiedades != nil        
        @propiedades.each { |propiedad| 
          num_casas = propiedad.casilla.num_casas
          num_hoteles = propiedad.casilla.num_hoteles
          props += "\t -" + propiedad.nombre + ":\t" + 
            (propiedad.hipotecada ? "Esta hipotecada" : "No esta hipotecada") + 
            "\n\t  Numero de Casas: " + num_casas.to_s + 
            "\tNumero de Hoteles: " + num_hoteles.to_s + "\n"
        }
      end
      
      @nombre.to_s + ": \n- Saldo:\t" + @saldo.to_s + "\n- Casilla: \n" + \
                @casilla_actual.to_s + "\n- Encarcelado:\t" + encarcelado.to_s + \
                "\n- Propiedades:\n" + props + "\n"
                    
    end

    #protected :pagar_impuestos    
    private :eliminar_de_mis_propiedades, :es_de_mi_propiedad, :tengo_saldo\
            , :cuantas_casas_hoteles_tengo
  end
  
end

File: 2GII/PDOO/QytetetJava/src/modeloqytetet/Jugador.java
package modeloqytetet;

import java.util.ArrayList;

public class Jugador {
    private Boolean encarcelado = false;
    private String  nombre;
    private int     saldo = 7500;
    
    private Sorpresa                   cartaLibertad = null;
    private ArrayList<TituloPropiedad> propiedades   = null;
    private Casilla                    casillaActual;
    static int                         factorEspeculador = 1;
    
    // ----------------------- Constructor -------------------------------------   
    public Jugador(String nombre) {
        this.nombre = nombre;
                
        // BUSCAR LA MANERA DE HACERLO SIN CREAR LA CASILLA 
        casillaActual = new OtraCasilla(0, TipoCasilla.SALIDA);
    }
    
    // NO FUNCIONA EN RUBY, AUNQUE SEAN DE LA MISMA CLASE NECESITA READER
    protected Jugador(Jugador jugador) {
        this.encarcelado   = jugador.getEncarcelado();
        this.nombre        = jugador.getNombre();
        this.saldo         = jugador.getSaldo();
        this.cartaLibertad = jugador.cartaLibertad;        
        
        // Intento de copia defensiva
        if (propiedades != null)
            propiedades = (ArrayList<TituloPropiedad>) jugador.getPropiedades().clone();        
        this.casillaActual = jugador.getCasillaActual();                
    }
    
    // ----------------------- Consultores -------------------------------------   
    public Casilla getCasillaActual() {
        return casillaActual;  
    }

    public String getNombre() {
        return nombre;
    }

    public int getSaldo() {
        return saldo;
    }        
    
    public Boolean getEncarcelado() {
        return encarcelado;
    }

    public ArrayList<TituloPropiedad> getPropiedades() {
        return propiedades;
    }        

    public int getFactorEspeculador() {
        return factorEspeculador;
    }        
    
    public int obtenerCapital() {
        int capital  = saldo;
        int nCasas,
            nHoteles;
        
        if (propiedades != null) {
            for (TituloPropiedad propiedad : propiedades) {
                nCasas = propiedad.getCasilla().getNumCasas();
                nHoteles = propiedad.getCasilla().getNumHoteles();
                capital += propiedad.getFactorRevalorizacion() * (nCasas + nHoteles)
                            + propiedad.getCasilla().getCoste();
            }
        }
        return capital;
    }
    
    public Boolean tengoPropiedades() {
        return propiedades != null;
    }    

    boolean tengoCartaLibertad() {
        return cartaLibertad != null;
    }    
    
    protected Boolean tengoSaldo(int cantidad) {
        return saldo >= cantidad;
    }    
    
    Boolean puedoEdificarCasa(Casilla casilla) {
        return esDeMiPropiedad(casilla) && tengoSaldo(casilla.getPrecioEdificar());
    }
    
    Boolean puedoEdificarHotel(Casilla casilla) {
        return esDeMiPropiedad(casilla) && tengoSaldo(casilla.getPrecioEdificar());
    }
    
    Boolean puedoHipotecar(Casilla casilla) {
        return !casilla.estaHipotecada();
    }
    
    Boolean puedoPagarHipoteca(Casilla casilla) {
        int precioDeshipotecar = casilla.getCosteCancelarHipoteca();
        
        return tengoSaldo(precioDeshipotecar);
    }
    
    Boolean puedoVenderPropiedad(Casilla casilla) {
        return esDeMiPropiedad(casilla) && !casilla.getTitulo().getHipotecada();
    }   
    
    private int cuantasCasasHotelesTengo()  {
        int total = 0;
        Casilla cas ;
        
        if (propiedades != null) {
            for (TituloPropiedad propiedad : propiedades) {
                cas = propiedad.getCasilla();
                total += cas.getNumCasas() + cas.getNumHoteles();
            }
        }
        
        return total;
    }    
    
    private Boolean esDeMiPropiedad(Casilla casilla) {
        return propiedades.contains(casilla.getTitulo());        
    }
    
    // ----------------------- Modificadores -----------------------------------   
    void modificarSaldo(int cantidad) {
        saldo += cantidad;
    }
    
    void setCartaLibertad(Sorpresa carta) {
        this.cartaLibertad = carta;
    }
    
    void setCasillaActual(Casilla casilla) {
        this.casillaActual = casilla;
    }
    
    void setEncarcelado(Boolean encarcelado) {
        this.encarcelado = encarcelado;
    }
    
    private void eliminarDeMisPropiedades(Casilla casilla) {
        propiedades.remove(casilla.getTitulo());
    }    
        
    // ----------------------- Métodos -----------------------------------------         
    protected Boolean actualizarPosicion(Casilla casilla) {
        Boolean tienePropietario = false;

        // Si la casilla objetivo es menor, es que ha pasado por la salida
        if (casilla.getNumeroCasilla() < casillaActual.getNumeroCasilla())
            modificarSaldo(Qytetet.SALDO_SALIDA);
                
        setCasillaActual(casilla);
        
        if (casilla.soyEdificable()) {
            tienePropietario = casilla.tengoPropietario();
            if (tienePropietario && !casilla.propietarioEncarcelado())  {
                int costeAlquiler = casilla.cobrarAlquiler();
                // En la otra clase se le ha añadido el dinero al propietario
                // Ahora se lo restamos al cliente
                modificarSaldo(-costeAlquiler);
            }                               
        }        
        else if (casilla.getTipo() == TipoCasilla.IMPUESTO)
            pagarImpuestos(casilla.getCoste());
        
        return tienePropietario;
    }
    
    Boolean comprarTitulo() {
        Boolean puedoComprar = false;
        
        if (casillaActual.soyEdificable() && !casillaActual.tengoPropietario()
                                          && casillaActual.precioTotalComprar() <= saldo) {            
            if (propiedades == null)
                propiedades = new ArrayList<TituloPropiedad>();                        
            
            propiedades.add(casillaActual.asignarPropietario(this));
            modificarSaldo(-casillaActual.getCoste());
            puedoComprar = true;
        }
                    
        return puedoComprar;
    }
    
    Sorpresa devolverCartaLibertad() {
        Sorpresa aDevolver = cartaLibertad;
        cartaLibertad = null;
        
        return aDevolver;
    }
    
    void irACarcel () {
        Casilla carcel = new OtraCasilla (15, TipoCasilla.CARCEL);
        irACarcel(carcel);
    }
    
    void irACarcel(Casilla casilla) {
        setCasillaActual(casilla);
        setEncarcelado(true);
    }
       
    ArrayList<TituloPropiedad> obtenerPropiedadesHipotecadas(Boolean hipotecada) {
        ArrayList<TituloPropiedad> propHipotecadas = null;
        
        for (TituloPropiedad propiedad : propiedades) {
            if (propiedad.getHipotecada() == hipotecada)
                propHipotecadas.add(propiedad) ;
        }
        
        return propHipotecadas;
    }
    
    void pagarCobrarPorCasaYHotel(int cantidad) {
        modificarSaldo(cantidad * cuantasCasasHotelesTengo());
    }
    
    Boolean pagarLibertad() {
        return pagarLibertad(Qytetet.PRECIO_LIBERTAD);
    }
    
    Boolean pagarLibertad(int cantidad) {
        Boolean pagado = false; 
        
        if (tengoSaldo(cantidad)) {
            modificarSaldo(-cantidad);
            pagado = true;
        }
        return pagado;
    }    
    
    void venderPropiedad(Casilla casilla) {
        int ganancia = casilla.venderTitulo();
        modificarSaldo(ganancia);
        
        eliminarDeMisPropiedades(casilla);
        if (propiedades.isEmpty())
            propiedades = null;
    }
    
    protected void pagarImpuestos(int cantidad) {
        modificarSaldo(-cantidad); 
    }    
    
    protected Especulador convertirme(int fianza) {
        return new Especulador(this, fianza);                
    }
        
    @Override
    public String toString() {
      String encar;
      String props = new String();
        if (encarcelado)
            encar = "Si";
        else 
            encar = "No";     
        
        int numCasas, numHoteles;
        if (propiedades != null) {
            for (TituloPropiedad propiedad : propiedades) {
                numCasas = propiedad.getCasilla().getNumCasas();
                numHoteles = propiedad.getCasilla().getNumHoteles();
                
                props += "\t -" + propiedad.getNombre() + ":\t" + 
                        (propiedad.getHipotecada() ? "Esta hipotecada" : 
                                                     "No esta hipotecada") + 
                        "\n\t  Numero de Casas: " + Integer.toString(numCasas) +
                        "\tNumero de Hoteles: " + Integer.toString(numHoteles) + "\n";
            }
        }
        
        /* Pre-Práctica 5
        return nombre + "\n- Saldo:\t" + saldo + "$\n- Casilla: \n" + 
                casillaActual + "\n- Encarcelado:\t" + encar + 
                "\n- Propiedades:\n" + props + "\n";
        */
        return nombre + "\n- Saldo:\t" + saldo + "\n- Encarcelado:\t" + encar + 
                "\n- Propiedades:\n" + props + "\n";
    }
}
